scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala

```diff
@@ -0,0 +1,61 @@
+package com.twitter.scalding.beam_backend
+
+import com.twitter.scalding.Config
+import com.twitter.scalding.beam_backend.BeamFunctions._
+import com.twitter.scalding.typed.TypedSource
+import org.apache.beam.sdk.Pipeline
+import org.apache.beam.sdk.transforms._
+import org.apache.beam.sdk.values.PCollection
+import scala.collection.JavaConverters._
+
+sealed abstract class BeamOp[+A] {
+  import BeamOp.TransformBeamOp
+
+  def run(pipeline: Pipeline): PCollection[_ <: A]
+
+  def map[C >: A, B](f: C => B)(implicit kryoCoder: KryoCoder): BeamOp[B] =
+    parDo(MapFn(f))
+
+  def parDo[C >: A, B](f: DoFn[C, B])(implicit kryoCoder: KryoCoder): BeamOp[B] = {
+    val pTransform = new PTransform[PCollection[C], PCollection[B]]() {
+      override def expand(input: PCollection[C]): PCollection[B] = input.apply(ParDo.of(f))
+    }
+    applyPTransform(pTransform)
+  }
+
+  def filter(f: A => Boolean)(implicit kryoCoder: KryoCoder): BeamOp[A] = {
+    val fn = ProcessFn(f.andThen(java.lang.Boolean.valueOf))
+    applyPTransform(Filter.by[A, ProcessFunction[A, java.lang.Boolean]](fn))
+  }
+
+  def applyPTransform[C >: A, B](f: PTransform[PCollection[C], PCollection[B]])(implicit kryoCoder: KryoCoder): BeamOp[B] =
+    TransformBeamOp(this, f, kryoCoder)
+
+  def flatMap[C >: A, B](f: C => TraversableOnce[B])(implicit kryoCoder: KryoCoder): BeamOp[B] =
```

Comment:
same comment here: `def flatMap[B](f: A => ...)` should work because `Function1` is contravariant in the first parameter.